### ISSUE TITLE
Message should consist of any number of any components that can be added/removed at will

### ISSUE THREAD

**Issue by smsm1** [cyclestreets/cyclescape#287]
If the attachment is able to have a thumbnail, then give it a thumbnail. The current behaviour is a bit broken as you can upload a file which doesn't have support for a thumbnail as an image, and it won't work as expected. e.g. http://www.cyclescape.org/threads/1180#message-13102

**Comment by mvl22** [comment]
https://twitter.com/MattTurnerSheff/status/693363637407342592

"Would really really like to be able to upload a gallery of photo's. It's a bit of an email barrage when you upload 5/6 images!"

**Comment by mvl22** [comment]
Facebook uses this model quite effectively, on mobile:

![0BB95083-F665-4E43-8043-AAF6B0DC5854](https://user-images.githubusercontent.com/361423/58674952-0a970a00-834a-11e9-8a61-1a442a0a797a.png)

**Comment by nikolai-b** [comment]
Sadly to keep the #769 we will need https://github.com/JangoSteve/remotipart because it isn't possible to send files via AJAX without hacking around.

**Comment by mvl22** [comment]
> it isn't possible to send files via AJAX without hacking around

I believe HTML5 Drag and drop Uploads work that way. So presumably this could switch to using that method when it's available.

**Comment by nikolai-b** [comment]
@mvl22 so I've made a start now I can create a post with multiple components:

![multiple component message](https://user-images.githubusercontent.com/6289830/65634570-690d8580-dfd6-11e9-93fd-2af8d097d8eb.png)

Obviously we could do with structuring it a bit better...

Now I assume that we want to combine all the emails into one so if someone posts a text, image and deadline then we'd send out one email:

A has posted a message:

'message'

and a photo which you can view at: link_to_photo

and ....


Also the same would be true if someone emailed in with multiple photos, we could just make a single message.

**Comment by mvl22** [comment]
> I assume that we want to combine all the emails into one

Yes, a message should be a collection of elements, whether that be posted/read via the web interface, or posted/read by e-mail.

**Comment by nikolai-b** [comment]
So creating Library items from messages is already a bit broken, the cog is a bit hidden and with the rich text messages the HTML appears in the library item body.  With multi- component messages then it becomes quite a bit more work to allow these to be converted to library items.

Can we remove this feature please?  The library needs quite a bit more work and I'm believe it is not worth the effort preserving the "Create Library Note" functionality.
![Screenshot from 2019-10-09 19-55-43](https://user-images.githubusercontent.com/6289830/66511460-ddfcb700-eace-11e9-81ff-16b1a8a89229.png)

If we do want to keep it then have you ideas as to what creating a library note from a message would do?  Just that the message text body (and fix the HTML issue)?

**Comment by nikolai-b** [comment]
@mvl22 any news on this?  Other than the "Create Library Note" feature https://github.com/cyclestreets/cyclescape/pull/878 is ready to be merged into staging and we can start more testing.

**Comment by nikolai-b** [comment]
@mvl22 any update?

**Comment by mvl22** [comment]
> So creating Library items from messages is already a bit broken, the cog is a bit hidden and with the rich text messages the HTML appears in the library item body. With multi- component messages then it becomes quite a bit more work to allow these to be converted to library items.
> 
> Can we remove this feature please? The library needs quite a bit more work and I'm believe it is not worth the effort preserving the "Create Library Note" functionality.

I don't want to remove the library feature as a whole, but for now I'm content with disabling the "creating Library items from messages" feature, so that we can unblock this other work.

> If we do want to keep it then have you ideas as to what creating a library note from a message would do? Just that the message text body (and fix the HTML issue)?

Yes, this is a good question and does need more thought. For instance, a deadline component within a message doesn't make sense in a Library context.

**Comment by nikolai-b** [comment]
@mvl22 this is now merged.  Would you mind commenting here when you've had a chance to check so we can deploy it to production.

**Comment by mvl22** [comment]
Bugs:

- CycleStreets Photo type: Caption is not copied in to the box after clicking "Select image".

- If you post a message, and then do a second message which contains a Photo component, that photo does not show a thumbnail. Similarly, CycleStreets photo panel does not load, and Map background does not show. I can see there is a Javascript warning appearing in the console, so there could be other side-effects here.

- Another second-posting bug: a "Street view messages caption can't be blank" red main warning flash always appears on submission.

-Ordering: Can we make the component order within the message the same as the tab ordering. For instance, I'm seeing the Text component after a Photo component.

**Comment by nikolai-b** [comment]
Thanks, I've done all these.  I couldn't recreate Javascript warnings so if you still get them would you mind posting here?

Thanks

**Comment by mvl22** [comment]
> I couldn't recreate Javascript warnings so if you still get them would you mind posting here?

The warnings only appear after doing a first posting. Did you try doing a posting, e.g. Text plus Map, and then seeing what errors happen upon trying to do a second posting?

**Comment by nikolai-b** [comment]
> Did you try doing a posting, e.g. Text plus Map, and then seeing what errors happen upon trying to do a second posting?

Yes.  If you wouldn't mind posting the JS errors here if they are still happening.  I believe I've fixed all the things you mentioned.

**Comment by nikolai-b** [comment]
@mvl22 let me know if this good to go now.  Also I received all the emails which look ok.

**Comment by mvl22** [comment]
Good to go now - thanks!

**Comment by nikolai-b** [comment]
Closing as this is live now

**Comment by mvl22** [comment]
Thanks, that's great work. I'll be sorting out the new UI for this shortly.
